In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("../data/HR_Analytics.csv")
print(f"shape: {df.shape}\n")
print(df.head(5))

shape: (1480, 38)

   EmpID  Age AgeGroup Attrition     BusinessTravel  DailyRate  \
0  RM297   18    18-25       Yes      Travel_Rarely        230   
1  RM302   18    18-25        No      Travel_Rarely        812   
2  RM458   18    18-25       Yes  Travel_Frequently       1306   
3  RM728   18    18-25        No         Non-Travel        287   
4  RM829   18    18-25       Yes         Non-Travel        247   

               Department  DistanceFromHome  Education EducationField  ...  \
0  Research & Development                 3          3  Life Sciences  ...   
1                   Sales                10          3        Medical  ...   
2                   Sales                 5          3      Marketing  ...   
3  Research & Development                 5          2  Life Sciences  ...   
4  Research & Development                 8          1        Medical  ...   

   RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0                         3             80         

In [3]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 1480 entries, 0 to 1479
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   EmpID                     1480 non-null   str    
 1   Age                       1480 non-null   int64  
 2   AgeGroup                  1480 non-null   str    
 3   Attrition                 1480 non-null   str    
 4   BusinessTravel            1480 non-null   str    
 5   DailyRate                 1480 non-null   int64  
 6   Department                1480 non-null   str    
 7   DistanceFromHome          1480 non-null   int64  
 8   Education                 1480 non-null   int64  
 9   EducationField            1480 non-null   str    
 10  EmployeeCount             1480 non-null   int64  
 11  EmployeeNumber            1480 non-null   int64  
 12  EnvironmentSatisfaction   1480 non-null   int64  
 13  Gender                    1480 non-null   str    
 14  HourlyRate         

In [4]:
print(df.describe())

               Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1480.000000  1480.000000       1480.000000  1480.000000         1480.0   
mean     36.917568   801.384459          9.220270     2.910811            1.0   
std       9.128559   403.126988          8.131201     1.023796            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   800.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1480.000000              1480.000000  1480.000000     1480.000000   
mean      1031.860811                 2.724324    65.845270        2.729730   
std        605.955046            

Target column: Attrition (categorical: 'Yes'/'No')

Key Numerical metrics: MonthlyIncome, Age, DistanceFromHome, YearsAtCompany, YearsWithCurrManager

Key Categorical metrics: Department, JobRole, AgeGroup, SalarySlab

Note: Notice that column 37 (YearsWithCurrManager) has 1423 non-null values out of 1480.

In [5]:
df_copy=df.copy()#safer copy

In [6]:
#KPI's
#total headcount and attrition rate
total_emp= len(df_copy)
attrition_count=(df_copy['Attrition']== 'Yes').sum()
attrition_rate=(attrition_count/ total_emp)* 100

print("==================================================")
print(f"TOTAL HEADCOUNT: {total_emp} employees")
print(f"TOTAL ATTRITION: {attrition_count} left the company")
print(f"OVERALL ATTRITION RATE: {attrition_rate:.2f}%")
print("==================================================\n")

TOTAL HEADCOUNT: 1480 employees
TOTAL ATTRITION: 238 left the company
OVERALL ATTRITION RATE: 16.08%



In [7]:
#financial overview
avg_income = df_copy['MonthlyIncome'].mean()
median_income = df_copy['MonthlyIncome'].median()
print(f"Average Monthly Income: ${avg_income:,.2f}")
print(f"Median Monthly Income: ${median_income:,.2f}\n")

Average Monthly Income: $6,504.99
Median Monthly Income: $4,933.00



In [8]:
#department breakdown
print("Department Performance Metrics:")
dept_analysis = df_copy.groupby('Department').agg(
    Total_Staff=('EmpID', 'count'),
    Avg_Monthly_Income=('MonthlyIncome', 'mean'),
    Departures=('Attrition', lambda x: (x == 'Yes').sum())
)
dept_analysis

Department Performance Metrics:


,Total_Staff,Avg_Monthly_Income,Departures
Department,,,
Human Resources,63,6654.507937,12
Research & Development,967,6280.365047,133
Sales,450,6966.737778,93


In [9]:
dept_analysis['Dept_Attrition_Rate_%'] = (dept_analysis['Departures'] / dept_analysis['Total_Staff']) * 100
print(dept_analysis.round(2))
print("\n")

                        Total_Staff  Avg_Monthly_Income  Departures  \
Department                                                            
Human Resources                  63             6654.51          12   
Research & Development          967             6280.37         133   
Sales                           450             6966.74          93   

                        Dept_Attrition_Rate_%  
Department                                     
Human Resources                         19.05  
Research & Development                  13.75  
Sales                                   20.67  




In [10]:
#experience and stability baseline
avg_tenure = df_copy['YearsAtCompany'].mean()
avg_total_experience = df_copy['TotalWorkingYears'].mean()
print(f"Average Loyalty: {avg_tenure:.1f} years at the company")
print(f"Average Industry Experience: {avg_total_experience:.1f} years\n")

Average Loyalty: 7.0 years at the company
Average Industry Experience: 11.3 years



In [11]:
#data integrity check
missing_managers = df_copy['YearsWithCurrManager'].isnull().sum()
print(f"Data Quality Red Flag: Missing values in 'YearsWithCurrManager': {missing_managers}")

Data Quality Red Flag: Missing values in 'YearsWithCurrManager': 57


In [13]:
#data cleaning
print(df_copy.isnull().sum())

EmpID                        0
Age                          0
AgeGroup                     0
Attrition                    0
BusinessTravel               0
DailyRate                    0
Department                   0
DistanceFromHome             0
Education                    0
EducationField               0
EmployeeCount                0
EmployeeNumber               0
EnvironmentSatisfaction      0
Gender                       0
HourlyRate                   0
JobInvolvement               0
JobLevel                     0
JobRole                      0
JobSatisfaction              0
MaritalStatus                0
MonthlyIncome                0
SalarySlab                   0
MonthlyRate                  0
NumCompaniesWorked           0
Over18                       0
OverTime                     0
PercentSalaryHike            0
PerformanceRating            0
RelationshipSatisfaction     0
StandardHours                0
StockOptionLevel             0
TotalWorkingYears            0
Training

In [16]:
#relplace null with median
df_copy['YearsWithCurrManager'] = df_copy['YearsWithCurrManager'].fillna(df_copy['YearsWithCurrManager'].median())

print("\nMissing Values AFTER Cleaning:")
print(df_copy.isnull().sum())


Missing Values AFTER Cleaning:
EmpID                       0
Age                         0
AgeGroup                    0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
SalarySlab                  0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
Training

In [17]:
#remove duplicates
before_drop = len(df_copy)
df_copy.drop_duplicates(inplace=True)
after_drop = len(df_copy)
print(f"\nDuplicates removed: {before_drop - after_drop}")


Duplicates removed: 7


In [ ]:
#relplace null with median
df_copy['YearsWithCurrManager'] = df_copy['YearsWithCurrManager'].fillna(df_copy['YearsWithCurrManager'].median())

print("\nMissing Values AFTER Cleaning:")
print(df_copy.isnull().sum())


Missing Values AFTER Cleaning:
EmpID                       0
Age                         0
AgeGroup                    0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
SalarySlab                  0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
Training

In [18]:
print("\nAttrition values:", df_copy['Attrition'].unique())
print("Department values:", df_copy['Department'].unique())


Attrition values: <StringArray>
['Yes', 'No']
Length: 2, dtype: str
Department values: <StringArray>
['Research & Development', 'Sales', 'Human Resources']
Length: 3, dtype: str


In [20]:
df_copy.to_csv('../data/HR_Analytics_Cleaned.csv', index=False)
print("\n Cleaned data saved to: data/HR_Analytics_Cleaned.csv")


 Cleaned data saved to: data/HR_Analytics_Cleaned.csv
